### ISSUE TITLE
libvfio-user context cannot be reused after client has disconnected

### ISSUE THREAD

**Issue by tmakatos** [nutanix/libvfio-user#392]
Related to https://github.com/spdk/spdk/issues/1813.

```
2021-03-11 06:35:22.008688] vfio_user.c:1330:vfio_user_log: *ERROR*: /var/run/muser/domain/muser0/8: end of file: Resource temporarily unavailable
[2021-03-11 06:35:22.008791] vfio_user.c: 257:fail_ctrlr: *ERROR*: :/var/run/muser/domain/muser0/8 failing controller
[2021-03-11 06:35:36.563177] vfio_user.c:1916:nvmf_vfio_user_poll_group_add: *DEBUG*: /var/run/muser/domain/muser0/8: add QP0=0x563ff2f3eed0(0x563ff2f3eed0) to poll_group=0x563ff2f45390
[2021-03-11 06:35:36.563204] vfio_user.c:1950:nvmf_vfio_user_poll_group_add: *DEBUG*: /var/run/muser/domain/muser0/8: sending connect fabrics command for QID=0 cntlid=0xffff
[2021-03-11 06:35:36.563430] vfio_user.c:1512:map_dma: *DEBUG*: /var/run/muser/domain/muser0/8: map IOVA 0-0xa0000
[2021-03-11 06:35:36.563443] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: added DMA region 0-0x9ffff offset=0 fd=503 prot=0x3
[2021-03-11 06:35:36.563497] vfio_user.c:1324:vfio_user_log: *INFO*: /var/run/muser/domain/muser0/8: new DMA region 0xc0000+0x20000 overlaps with DMA region 0xc0000-0xb000
[2021-03-11 06:35:36.563536] vfio_user.c:1324:vfio_user_log: *INFO*: /var/run/muser/domain/muser0/8: failed to add DMA region 0xc0000-0xdffff offset=0 fd=-1: Input/output error
[2021-03-11 06:35:36.563541] vfio_user.c:1330:vfio_user_log: *ERROR*: /var/run/muser/domain/muser0/8: msg0x2: cmd 2 failed: Input/output error
[2021-03-11 06:35:36.563632] vfio_user.c:1324:vfio_user_log: *INFO*: /var/run/muser/domain/muser0/8: new DMA region 0xe0000+0x20000 overlaps with DMA region 0xec000-0x4000
[2021-03-11 06:35:36.563664] vfio_user.c:1324:vfio_user_log: *INFO*: /var/run/muser/domain/muser0/8: failed to add DMA region 0xe0000-0xfffff offset=0 fd=-1: No such process
[2021-03-11 06:35:36.563668] vfio_user.c:1330:vfio_user_log: *ERROR*: /var/run/muser/domain/muser0/8: msg0x3: cmd 2 failed: No such process
[2021-03-11 06:35:36.563758] vfio_user.c:1512:map_dma: *DEBUG*: /var/run/muser/domain/muser0/8: map IOVA 0x100000-0x3ff00000
[2021-03-11 06:35:36.563766] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: added DMA region 0x100000-0x3fffffff offset=0x100000 fd=504 prot=0x3
[2021-03-11 06:35:36.563839] vfio_user.c:1512:map_dma: *DEBUG*: /var/run/muser/domain/muser0/8: map IOVA 0xfffc0000-0x40000
[2021-03-11 06:35:36.563846] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: added DMA region 0xfffc0000-0xffffffff offset=0 fd=-1 prot=0x1
[2021-03-11 06:35:36.564072] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: devinfo flags 0x3, num_regions 10, num_irqs 5
[2021-03-11 06:35:36.564151] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: region_info[0] offset 0 flags 0xf size 16384 argsz 64
[2021-03-11 06:35:36.564207] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: dev_get_caps: area 0 [0x1000-0x2000)
[2021-03-11 06:35:36.564214] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: region_info[0] offset 0 flags 0xf size 16384 argsz 64
[2021-03-11 06:35:36.564271] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: region_info[1] offset 0x10000000000 flags 0 size 0 argsz 32
[2021-03-11 06:35:36.564315] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: region_info[2] offset 0x20000000000 flags 0 size 0 argsz 32
[2021-03-11 06:35:36.564350] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: region_info[3] offset 0x30000000000 flags 0 size 0 argsz 32
[2021-03-11 06:35:36.564379] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: region_info[4] offset 0x40000000000 flags 0x3 size 4096 argsz 32
[2021-03-11 06:35:36.564425] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: region_info[5] offset 0x50000000000 flags 0x3 size 4096 argsz 32
[2021-03-11 06:35:36.564470] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: region_info[7] offset 0x70000000000 flags 0x3 size 4096 argsz 32
[2021-03-11 06:35:36.564667] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: R 7 0-0x1000
[2021-03-11 06:35:36.566619] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: setting IRQ INTx flags=0x24
[2021-03-11 06:35:36.566666] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: event fd[0]=505
[2021-03-11 06:35:36.566911] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: setting IRQ ERR flags=0x24
[2021-03-11 06:35:36.566935] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: event fd[0]=502
[2021-03-11 06:35:36.572335] vfio_user.c:1326:vfio_user_log: *NOTICE*: /var/run/muser/domain/muser0/8: ignore IRQ DATA_NONE, action=0x20, index=0, start=0, count=0
[2021-03-11 06:35:36.573000] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: R 7 0x44-0x46
[2021-03-11 06:35:36.573079] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: R 7 0x4-0x6
[2021-03-11 06:35:36.573143] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: W 7 0x4-0x6
[2021-03-11 06:35:36.573164] vfio_user.c:1324:vfio_user_log: *INFO*: /var/run/muser/domain/muser0/8: I/O space disabled
[2021-03-11 06:35:36.573181] vfio_user.c:1324:vfio_user_log: *INFO*: /var/run/muser/domain/muser0/8: memory space disabled
[2021-03-11 06:35:36.573194] vfio_user.c:1324:vfio_user_log: *INFO*: /var/run/muser/domain/muser0/8: bus master disabled
[2021-03-11 06:35:36.573210] vfio_user.c:1324:vfio_user_log: *INFO*: /var/run/muser/domain/muser0/8: INTx emulation enabled
[2021-03-11 06:35:36.573230] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: Device reset called by client
[2021-03-11 06:35:36.573362] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: setting IRQ INTx flags=0x24
[2021-03-11 06:35:36.573396] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: event fd[0]=505
[2021-03-11 06:35:36.573464] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: W 7 0x10-0x14
[2021-03-11 06:35:36.573484] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: BAR0 addr 0x0
[2021-03-11 06:35:36.573506] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: W 7 0x14-0x18
[2021-03-11 06:35:36.573551] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: W 7 0x18-0x1c
[2021-03-11 06:35:36.573576] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: W 7 0x1c-0x20
[2021-03-11 06:35:36.573599] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: W 7 0x20-0x24
[2021-03-11 06:35:36.573618] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: BAR4 addr 0x0
[2021-03-11 06:35:36.573641] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: W 7 0x24-0x28
[2021-03-11 06:35:36.573660] vfio_user.c:1322:vfio_user_log: *DEBUG*: /var/run/muser/domain/muser0/8: BAR5 addr 0x0
[2021-03-11 06:35:36.580954] vfio_user.c:1324:vfio_user_log: *INFO*: /var/run/muser/domain/muser0/8: failed to remove DMA region 0xc0000-0xdffff: No such file or directory
[2021-03-11 06:35:36.580993] vfio_user.c:1330:vfio_user_log: *ERROR*: /var/run/muser/domain/muser0/8: msg0x1f: cmd 3 failed: No such file or directory
[2021-03-11 06:35:36.581212] vfio_user.c:1324:vfio_user_log: *INFO*: /var/run/muser/domain/muser0/8: failed to remove DMA region 0xe0000-0xfffff: No such file or directory
[2021-03-11 06:35:36.581244] vfio_user.c:1330:vfio_user_log: *ERROR*: /var/run/muser/domain/muser0/8: msg0x20: cmd 3 failed: No such file or directory
[2021-03-11 06:35:36.581344] vfio_user.c:1569:unmap_dma: *DEBUG*: /var/run/muser/domain/muser0/8: unmap IOVA 0x100000
nvmf_tgt: /root/src/libvfio-user/lib/dma.c:169: dma_controller_remove_region: Assertion `region->refcnt == 0' failed.
```

Looks like we need to clean up the libvfio-user context when the client disconnects.

**Comment by changpe1** [comment]
when connection disconnected, libvfio-user can provide an API like vfu_reset() to release dma region resources, maybe reset some PCI registers.

**Comment by jlevon** [comment]
Fixed by #419
